domain/schema/secret.go

```diff
@@ -0,0 +1,282 @@
+        CONSTRAINT fk_secret_model_owned_secret_metadata_uuid
+            FOREIGN KEY (secret_metadata_uuid)
+            REFERENCES secret_metadata (uuid)
+    );
+
+CREATE UNIQUE INDEX idx_secret_model_owned_label ON secret_model_owned (label);
+
+-- It's ok to duplicate this table for each model database.
+CREATE TABLE
+    rotate_policy (
+        id INT PRIMARY KEY,
+        policy TEXT NOT NULL,
+        CONSTRAINT uniq_rotate_policy_policy
+            UNIQUE(policy)
+    );
+
+INSERT INTO rotate_policy VALUES
+    (0, 'never'),
+    (1, 'hourly'),
+    (2, 'daily'),
+    (3, 'weekly'),
+    (4, 'monthly'),
+    (5, 'quarterly'),
+    (6, 'yearly');
+
+CREATE TABLE
+    secret_metadata (
+        uuid TEXT PRIMARY KEY,
+        version INT,
+        description TEXT,
+        latest_revision INT NOT NULL,
+        latest_expire_time DATETIME,
+        rotate_policy TEXT,
+        auto_prune BOOLEAN NOT NULL DEFAULT (FALSE),
+        create_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        update_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        CONSTRAINT fk_secret_metadata_rotate_policy
+            FOREIGN KEY (rotate_policy)
+            REFERENCES rotate_policy (policy)
+    );
+
+CREATE TABLE
+    secret_value_ref (
+        uuid TEXT PRIMARY KEY,
+        backend_id TEXT NOT NULL,
+        revision_id TEXT NOT NULL
+    );
+
+CREATE TABLE
+    secret_revision (
+        metadata_uuid TEXT NOT NULL,
+        revision INT NOT NULL,
+        data BLOB,
+        value_ref_uuid TEXT,
+        obsolete BOOLEAN NOT NULL DEFAULT (FALSE), -- TODO: do we still need this?
+        pending_delete BOOLEAN NOT NULL DEFAULT (FALSE),
+        expire_time DATETIME,
+        create_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        update_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        CONSTRAINT pk_secret_revision_metadata_uuid_revision
+            PRIMARY KEY (metadata_uuid, revision),
+        CONSTRAINT fk_secret_revision_secret_metadata_uuid 
+            FOREIGN KEY (metadata_uuid)
+            REFERENCES secret_metadata (uuid),
+        CONSTRAINT fk_secret_revision_secret_value_ref_uuid
+            FOREIGN KEY (value_ref_uuid)
+            REFERENCES secret_value_ref (uuid),
+        CONSTRAINT secret_revision_data_or_value_ref_not_null
+            CHECK (
+                (data IS NOT NULL AND value_ref_uuid IS NULL) OR
+                (data IS NULL AND value_ref_uuid IS NOT NULL)
+            )
+    );
+
+CREATE TABLE
+    secret_application_consumed (
+        secret_metadata_uuid TEXT NOT NULL,
+        application_uuid TEXT NOT NULL,
+        label TEXT,
+        current_revision INT NOT NULL,
+        -- TODO:latest_revision was for the consumer watcher can be 
+        -- triggered when a new secret revision is added.
+        -- ????
+        latest_revision INT NOT NULL,
```

Comment:
We can now avoid this denormalisation (here and elsewhere). We need to understand all the required watchers and what would trigger them; then we can look at inserting a row in the change event table as part of the txn to update the data model rather than using triggers if it's something complex, or where appropriate, we can just  use triggers.